rfcs/0000-client-block-filter/0000-client-block-filter.md

```diff
@@ -0,0 +1,136 @@
+
+#### GetBlockFilterHashes
+`GetBlockFilterHashes` is used to request verifiable filter hashes for a particular range of blocks:
+
+```
+struct GetBlockFilterHashes {
+    start_number:   Uint64,  // The height of the first block in the requested range
+}
+```
+
+#### BlockFilterHashes
+`BlockFilterHashes` is sent in response to `GetBlockFilterHashes`:
+
+```
+table BlockFilterHashes {
+    start_number:               Uint64,     // The height of the first block in the requested range
+    parent_block_filter_hash:   Byte32,     // The hash of the parent block filter
+    block_filter_hashes:        Byte32Vec,  // The hashes of the block filters in the range
+}
+```
+
+1. The `start_number` SHOULD match the field in the GetBlockFilterHashes request.
+2. The `block_filter_hashes` field size SHOULD not be larger than 2000.
+
+
+#### GetBlockFilterCheckPoints
+`GetBlockFilterCheckPoints` is used to request filter hashes at evenly spaced intervals over a range of blocks. Clients may use filter hashes from `GetBlockFilterHashes` to connect these checkpoints, as is described in the
+[Client Operation](#client-operation) section below:
+
+```
+struct GetBlockFilterCheckPoints {
+    start_number:   Uint64,     // The height of the first block in the requested range
+}
+```
+
+#### BlockFilterCheckPoints
+`BlockFilterCheckPoints` is sent in response to `GetBlockFilterCheckPoints`. The filter hashes included are the set of all filter hashes on the requested blocks range where the height is a multiple of the interval 2000:
+
+```
+table BlockFilterCheckPoints {
+    start_number:           Uint64,
+    block_filter_hashes:    Byte32Vec,
+}
+```
+
+1. The `start_number` SHOULD match the field in the GetBlockFilterCheckPoints request.
+2. The `block_filter_hashes` field size should not be larger than 2000.
+
+### Filter Data Generation
+
+We follow the BIP158 for filter data generation and use the same Golomb-Coded Sets parameters P and M values. The only difference is that we only use cell's lock/type script hash as the filter data:
+
+```
+filter.add_element(cell.lock.calc_script_hash().as_slice());
+if let Some(type_script) = cell.type_().to_opt() {
+    filter.add_element(type_script.calc_script_hash().as_slice());
+}
+```
+
+### Node Operation
+
+Full nodes MAY opt to support this RFC, such nodes SHOULD treat the filters as an additional index of the blockchain. For each new block that is connected to the main chain, nodes SHOULD generate filters and persist them. Nodes that are missing filters and are already synced with the blockchain SHOULD reindex the chain upon start-up, constructing filters for each block from genesis to the current tip.
+
+Nodes SHOULD NOT generate filters dynamically on request, as malicious peers may be able to perform DoS attacks by requesting small filters derived from large blocks. This would require an asymmetical amount of I/O on the node to compute and serve.
```

Comment:
```suggestion
Nodes SHOULD NOT generate filters dynamically on request, as malicious peers may be able to perform DoS attacks by requesting small filters derived from large blocks. This would require an asymmetrical amount of I/O on the node to compute and serve.
```